lib/streamlit/state/session_state.py

```diff
@@ -278,6 +304,62 @@ def _missing_key_error_message(key: str) -> str:
     )
 
 
+class WidgetStateReport(Protocol[T_co]):
+    """Should be usable by widget code to determine what value to return, and
+    whether to update the UI.
+    """
+
+    @property
+    @abstractmethod
+    def return_value(self) -> T_co:
+        ...
+
+    @property
+    @abstractmethod
+    def set_frontend_value(self) -> bool:
+        ...
```

Comment:
Let's document the property values as well:

```python
class RegisterWidgetResult:
    """Result returned by `register_widget`."""

    @property
    @abstractmethod
    def return_value(self) -> T_co:
        """The widget's current value."""
        ...

    @property
    @abstractmethod
    def set_frontend_value(self) -> bool:
        """True if the widget's value is different from the value most
        recently returned from the frontend.
        """
        ...
```

Reply:
Documentation added.